core/chains/evm/txm/txm.go

```diff
@@ -0,0 +1,462 @@
+	for {
+		_, unconfirmedCount, err := t.txStore.FetchUnconfirmedTransactionAtNonceWithCount(ctx, 0, address)
+		if err != nil {
+			return false, err
+		}
+
+		// Optimistically send up to maxInFlightSubset of the maxInFlightTransactions. After that threshold, broadcast more cautiously
+		// by checking the pending nonce so no more than maxInFlightSubset can get stuck simultaneously i.e. due
+		// to insufficient balance. We're making this trade-off to avoid storing stuck transactions and making unnecessary
+		// RPC calls. The upper limit is always maxInFlightTransactions regardless of the pending nonce.
+		if unconfirmedCount >= maxInFlightSubset {
+			if unconfirmedCount > maxInFlightTransactions {
+				t.lggr.Warnf("Reached transaction limit: %d for unconfirmed transactions", maxInFlightTransactions)
+				return true, nil
+			}
+			pendingNonce, e := t.client.PendingNonceAt(ctx, address)
+			if e != nil {
+				return false, e
+			}
+			nonce := t.getNonce(address)
+			if nonce > pendingNonce {
+				t.lggr.Warnf("Reached transaction limit. LocalNonce: %d, PendingNonce %d, unconfirmedCount: %d",
+					nonce, pendingNonce, unconfirmedCount)
+				return true, nil
+			}
+		}
+
+		nonce := t.getNonce(address)
+		tx, err := t.txStore.UpdateUnstartedTransactionWithNonce(ctx, address, nonce)
+		if err != nil {
+			return false, err
+		}
+		if tx == nil {
+			return false, nil
+		}
+		t.setNonce(address, nonce+1)
+
+		if err := t.createAndSendAttempt(ctx, tx, address); err != nil {
+			return true, err
+		}
+	}
+}
+
+func (t *Txm) createAndSendAttempt(ctx context.Context, tx *types.Transaction, address common.Address) error {
+	attempt, err := t.attemptBuilder.NewAttempt(ctx, t.lggr, tx, t.config.EIP1559)
+	if err != nil {
+		return err
+	}
+
+	if tx.Nonce == nil {
+		return fmt.Errorf("nonce for txID: %v is empty", tx.ID)
+	}
+	if err = t.txStore.AppendAttemptToTransaction(ctx, *tx.Nonce, address, attempt); err != nil {
+		return err
+	}
+
+	return t.sendTransactionWithError(ctx, tx, attempt, address)
+}
+
+func (t *Txm) sendTransactionWithError(ctx context.Context, tx *types.Transaction, attempt *types.Attempt, address common.Address) (err error) {
+	if tx.Nonce == nil {
+		return fmt.Errorf("nonce for txID: %v is empty", tx.ID)
+	}
+	start := time.Now()
+	txErr := t.client.SendTransaction(ctx, tx, attempt)
+	tx.AttemptCount++
+	t.lggr.Infow("Broadcasted attempt", "tx", tx, "attempt", attempt, "duration", time.Since(start), "txErr: ", txErr)
+	if txErr != nil && t.errorHandler != nil {
+		if err = t.errorHandler.HandleError(tx, txErr, t.attemptBuilder, t.client, t.txStore, t.setNonce, false); err != nil {
+			return
+		}
+	} else if txErr != nil {
+		pendingNonce, err := t.client.PendingNonceAt(ctx, address)
+		if err != nil {
+			return err
+		}
+		if pendingNonce <= *tx.Nonce {
+			t.lggr.Debugf("Pending nonce for txID: %v didn't increase. PendingNonce: %d, TxNonce: %d", tx.ID, pendingNonce, *tx.Nonce)
```

Comment:
So here's the part where we don't look at the error reason right?
What happens if:
- RPC says funds not enough
- RPC says gas limit too low
- RPC says Tx is invalid for censored
- RPC says Tx causes Overflow

This was my main concern for not reading the error responses. How do we decide further action if we don't know why the tx failed?

Reply:
Is the concern around logging or taking action? All of these errors are being logged. If we want improved logging we can always implement the same error parsing we have now. But in terms of functionality,  pretty much every error that doesn't get fixed via a retry will make the node get stuck no matter if it's TXM v1 or v2. With TXMv2, after a few retries, you'll get an error message and notice that the node got stuck. To me, it seems unreasonable to have a dependency on something that is slowing integration time and it's very unreliable just so we can potentially notice an issue a few txs earlier for a node that eventually will get stuck anyway.